sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonSquidSensor.java

```diff
@@ -171,43 +173,35 @@ private PythonConfiguration createConfiguration() {
     return new PythonConfiguration(context.fileSystem().encoding());
   }
 
-  private void save(Collection<SourceCode> squidSourceFiles) {
-    for (SourceCode squidSourceFile : squidSourceFiles) {
-      SourceFile squidFile = (SourceFile) squidSourceFile;
-
-      InputFile inputFile = context.fileSystem().inputFile(context.fileSystem().predicates().is(new java.io.File(squidFile.getKey())));
+  private void saveMeasures(InputFile inputFile, SensorContext sensorContext, MetricVisitor metricVisitor, PythonVisitorContext visitorContext) {
+    metricVisitor.scanFile(visitorContext);
+    FileLinesVisitor fileLinesVisitor = metricVisitor.fileLinesVisitor();
 
-      noSonarFilter.noSonarInFile(inputFile, squidFile.getNoSonarTagLines());
+    noSonarFilter.noSonarInFile(inputFile, fileLinesVisitor.getLinesWithNoSonar());
 
-      saveFilesComplexityDistribution(inputFile, squidFile);
-      saveFunctionsComplexityDistribution(inputFile, squidFile);
-      saveMeasures(inputFile, squidFile);
-      saveIssues(inputFile, squidFile);
-    }
-  }
+    saveFilesComplexityDistribution(inputFile, metricVisitor);
+    saveFunctionsComplexityDistribution(inputFile, metricVisitor);
 
-  private void saveMeasures(InputFile inputFile, SourceFile squidFile) {
-    saveMetricOnFile(inputFile, CoreMetrics.NCLOC, squidFile.getInt(PythonMetric.LINES_OF_CODE));
-    saveMetricOnFile(inputFile, CoreMetrics.STATEMENTS, squidFile.getInt(PythonMetric.STATEMENTS));
-    saveMetricOnFile(inputFile, CoreMetrics.FUNCTIONS, squidFile.getInt(PythonMetric.FUNCTIONS));
-    saveMetricOnFile(inputFile, CoreMetrics.CLASSES, squidFile.getInt(PythonMetric.CLASSES));
-    saveMetricOnFile(inputFile, CoreMetrics.COMPLEXITY, squidFile.getInt(PythonMetric.COMPLEXITY));
-    saveMetricOnFile(inputFile, CoreMetrics.COMMENT_LINES, squidFile.getInt(PythonMetric.COMMENT_LINES));
+    saveMetricOnFile(inputFile, CoreMetrics.NCLOC, fileLinesVisitor.getLinesOfCode().size());
+    saveMetricOnFile(inputFile, CoreMetrics.STATEMENTS, metricVisitor.numberOfStatements());
+    saveMetricOnFile(inputFile, CoreMetrics.FUNCTIONS, metricVisitor.numberOfFunctions());
+    saveMetricOnFile(inputFile, CoreMetrics.CLASSES, metricVisitor.numberOfClasses());
+    saveMetricOnFile(inputFile, CoreMetrics.COMPLEXITY, metricVisitor.complexity());
+    saveMetricOnFile(inputFile, CoreMetrics.COMMENT_LINES, fileLinesVisitor.getLinesOfComments().size());
   }
 
-  private <T extends Serializable> void saveMetricOnFile(InputFile inputFile, Metric metric, T value) {
-    context.<T>newMeasure()
+  private <T extends Serializable> void saveMetricOnFile(InputFile inputFile, Metric<Integer> metric, Integer value) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') T is not used in the method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2326)